python/ray/autoscaler/_private/_azure/node_provider.py

```diff
@@ -302,67 +332,131 @@ def terminate_node(self, node_id):
+        )
+        for n in nics:
             try:
-                # delete nic
-                delete = get_azure_sdk_function(
-                    client=self.network_client.network_interfaces,
-                    function_name="delete",
-                )
-                delete(
-                    resource_group_name=resource_group,
-                    network_interface_name=metadata["nic_name"],
+                nic_lros.append(
+                    delete(
+                        resource_group_name=resource_group,
+                        network_interface_name=n,
+                    )
                 )
             except Exception as e:
-                logger.warning("Failed to delete nic: {}".format(e))
-
-            # delete ip address
-            if "public_ip_name" in metadata:
-                try:
-                    delete = get_azure_sdk_function(
-                        client=self.network_client.public_ip_addresses,
-                        function_name="delete",
-                    )
+                logger.warning("Failed to delete NIC: {}".format(e))
+
+        while not all(nlro.done() for nlro in nic_lros):
+            time.sleep(0.1)
+
+        # Delete Public IPs
+        delete = get_azure_sdk_function(
+            client=self.network_client.public_ip_addresses,
+            function_name="delete",
+        )
+        ip_lros = []
+        for ip in ips:
+            try:
+                ip_lros.append(
                     delete(
                         resource_group_name=resource_group,
-                        public_ip_address_name=metadata["public_ip_name"],
+                        public_ip_address_name=ip,
                     )
-                except Exception as e:
-                    logger.warning("Failed to delete public ip: {}".format(e))
-
-            # delete disks
-            for disk in disks:
-                try:
-                    delete = get_azure_sdk_function(
-                        client=self.compute_client.disks, function_name="delete"
-                    )
-                    delete(resource_group_name=resource_group, disk_name=disk)
-                except Exception as e:
-                    logger.warning("Failed to delete disk: {}".format(e))
+                )
+            except Exception as e:
+                logger.warning("Failed to delete public IP: {}".format(e))
+
+        while not all(dlro.done() for dlro in disk_lros):
+            time.sleep(0.1)
+        while not all(iplro.done() for iplro in ip_lros):
+            time.sleep(0.1)
 
     def _get_node(self, node_id):
         self._get_filtered_nodes({})  # Side effect: updates cache
         return self.cached_nodes[node_id]
 
     def _get_cached_node(self, node_id):
-        if node_id in self.cached_nodes:
-            return self.cached_nodes[node_id]
-        return self._get_node(node_id=node_id)
+        return self.cached_nodes.get(node_id) or self._get_node(node_id=node_id)
```

Comment:
```suggestion
        return self.cached_nodes.get(node_id, self._get_node(node_id=node_id))
```

Reply:
@gramhagen I actually had to revert this suggestion as it results in invalidating the cache every time (since `self._get_node(node_id=node_id)` is always called). The `or` allows for short-circuiting that call.

Reply:
oops, good catch!